### PR TITLE
Update Ingress spec

### LINKED ISSUE #46: Validate Ingress
- [x] Make sure crt & key have proper pem format.
 - [x] Multiple hosts using same port for TCP ingress

### LINKED ISSUE #73: Run L7 ingress on non-standard ports
Test case:
- HTTP on ports other than 80 & 443.
- 443: terminate SSL, 80: redirects to 443. (noSSL)
- HTTP (noSSL on port 443), application itself terminates SSL.

### LINKED ISSUE #128: Allow users to specify NodePort for service ports in NodePort mode.
Test case:
- Initially set.
- Initially not set, then set via update.

TCP and HTTP

### LINKED ISSUE #138: Correctly compute content hash for HAproxy config
Scenarios to handle:

- [ ] User changes the order of hosts in Ingress YAML.
- [ ] Skip blank lines.
- [ ] `TrimRight` generated config before calculating hash.

### LINKED ISSUE #271: Improve Prometheus labels from HAProxy Exporter
Currently, the backend label applied to the HAProxy exporter metrics with the HAProxy backend doesn't identify the Ingress.

```
haproxy_backend_current_session_rate{backend="http-service-zzt4r5"} 0
```

Could we change the name of the `backend` to be prefixed by the Ingress name?

For example, for an Ingress with `metadata.name = my-app`:

```
haproxy_backend_current_session_rate{backend="my-app-http-service-zzt4r5"} 0
```

This would allow for filtering by Ingress in Prometheus. Alternatively, a new label could be added, but I don't know if HAProxy Exporter or the Prometheus ServiceMonitor will allow adding arbitrary static labels.

### LINKED ISSUE #310: haproxy stats, named services
I'm splitting up the issue #303 , in this I'm requesting if it's possible to better name service while creating the haproxy.cfg

because : 
haproxy stats give stats about each services.. which is good but the names are all like :
http-service-cm4ufh
http-service-uykjxd

in a page with 30 of them.. the mess is guarantee.. there is the possibility to add a name?, instead of
http-service-$HASH , http-service-$SERVICENAME-$HASH ? I suppose is doable without big effort :)

Thank you for your time,
Francesco

### LINKED ISSUE #318: Show validation error if multiple TCP rules are sharing the same port

### LINKED ISSUE #319: Revise TCP secret name
- Take secret name from TLS list.
- Add NoSSL to avoid TLS selectively.

### LINKED ISSUE #333: Open port 443 in HTTP mode
From "ronenf" in our slack channel:

Hi I am having problems getting my app running on port 443. Here part of the ingress:
```
- host: cvc-dev.vatit.io
   http:
     paths:
     - path: /
       backend:
         serviceName: vatit-cvc-dev.vatit-cvc-dev
         servicePort: 443
```

heres the service:
```
spec:
 selector:
     app: vatit-cvc-dev
 ports:
   - port: 443
     targetPort: 443
     name: https
   - port: 80
     targetPort: 80
     name: http
```

The cert is built into the app. I can hit the app on port 80 (It just returns an error saying i cant comunicate https over http). With https infront of the url i cant hit it.

### LINKED ISSUE #346: Validate existing Ingress before starting operator
- []TCP -> TCP
- HTTP xor TCP

### PR DESCRIPTION
**Pull request by tamalsaha**
Fixes #46
Fixes #73
Fixes #128
Fixes #310
Fixes #138
Fixes #319
Fixes #271
Fixes #333
Fixes #346
Fixes #318

### CODE REVIEW

**Comment by tamalsaha** [comment]
@sadlil , it's ready for review.

File: pkg/haproxy/types.go [hunk 1]
```diff
@@ -0,0 +1,102 @@
...
+	RewriteRules []string
+	// Deprecated
+	HeaderRules []string
+	Endpoints   []*Endpoint
+}
+
+func (be *Backend) canonicalize() {
+	sort.Strings(be.BackendRules)
+	sort.Strings(be.RewriteRules)
+	sort.Strings(be.HeaderRules)
```
Comment (sadlil) [inline pkg/haproxy/types.go]: Are these rules going to be sort lexicographically? If yes. then it is wrong.
These rules have meaning in the given order.

Comment (tamalsaha) [inline pkg/haproxy/types.go]: No. Reverted this 3 lines.

File: api/validator.go [hunk 2]
```diff
@@ -0,0 +1,143 @@
...
+	return fmt.Sprintf("%s:%d", a.Host, a.Port)
+}
+
+func (r Ingress) IsValid(cloudProvider string) error {
+	addrs := make(map[address]int)
+	nodePorts := make(map[int]int)
+	for ri, rule := range r.Spec.Rules {
+		if rule.HTTP != nil {
+			addr := address{Host: rule.Host}
+			if port, err := checkRequiredPort(rule.HTTP.Port); err != nil {
```
Comment (sadlil) [inline api/validator.go]: what if i do not set any port value. this should set 80/443 by default for http, as it currently is. this should not through an error.

File: api/validator.go [hunk 3]
```diff
@@ -0,0 +1,143 @@
...
+				return fmt.Errorf("spec.rule[%d].http.port %s is invalid. Reason: %s", ri, rule.HTTP.Port, err)
+			} else {
+				addr.Port = port
+			}
+			if ei, found := addrs[addr]; found {
```
Comment (sadlil) [inline api/validator.go]: i thik it is possible to reuse same host:port in two different rule in http mode.

File: api/validator.go [hunk 4]
```diff
@@ -0,0 +1,143 @@
...
+			if rule.TCP.Backend.ServiceName == "" {
+				return fmt.Errorf("spec.rule[%d].tcp is missing serviceName for addr %s", ri, addr)
+			}
+			if _, err := checkRequiredPort(rule.TCP.Backend.ServicePort); err != nil {
+				return fmt.Errorf("spec.rule[%d].tcp is using invalid servicePort %s for addr %s. Reason: %s", ri, rule.TCP.Backend.ServicePort, addr, err)
+			}
+			if len(rule.TCP.Backend.HeaderRule) > 0 {
+				return fmt.Errorf("spec.rule[%d].tcp.backend.headerRule must be empty for addr %s", ri, addr)
+			}
+			if len(rule.TCP.Backend.RewriteRule) > 0 {
```
Comment (sadlil) [inline api/validator.go]: what if we do not allow this field to be here for tcp, as:.
```
// IngressBackend describes all endpoints for a given service and port.
type IngressBackend struct {
	// Specifies the name of the referenced service.
	ServiceName string `json:"serviceName,omitempty"`

	// Specifies the port of the referenced service.
	ServicePort intstr.IntOrString `json:"servicePort,omitempty"`
}

type HTTPIngressBackend struct {
        IngressBackend `json:",inline"`

	HostNames []string `json:"hostNames,omitempty"`

	// Serialized HAProxy rules to apply on server backend including
	// request, response or header rewrite. acls also can be used.
	// https://cbonte.github.io/haproxy-dconv/1.7/configuration.html#1
	BackendRule []string `json:"backendRule,omitempty"`

	// Path rewrite rules with haproxy formatted regex.
	//
	// Deprecated: Use backendRule, will be removed.
	RewriteRule []string `json:"rewriteRule,omitempty"`

	// Header rules to modifies the header.
	//
	// Deprecated: Use backendRule, will be removed.
	HeaderRule []string `json:"headerRule,omitempty"`
}
```

File: api/validator.go [hunk 5]
```diff
@@ -0,0 +1,143 @@
...
+			} else {
+				addr.Port = port
+			}
+			if ei, found := addrs[addr]; found {
+				return fmt.Errorf("spec.rule[%d].http is reusing addr %s, also used in spec.rule[%d]", ri, addr, ei)
+			}
+			addrs[addr] = ri
+
+			if np, err := checkOptionalPort(rule.HTTP.NodePort); err != nil {
+				return fmt.Errorf("spec.rule[%d].http.nodePort %s is invalid. Reason: %s", ri, rule.HTTP.NodePort, err)
```
Comment (sadlil) [inline api/validator.go]: if we through here optional Port is not really optional. If i do not set this valu in yaml it must set system-default.

File: api/validator.go (hunk 3 above)
Comment (tamalsaha) [inline api/validator.go]: That is intended. Our TemplateData parser does not handle this case. So, no reason to support it here. Is there any reason why someone will want to do so?

File: api/validator.go (hunk 5 above)
Comment (tamalsaha) [inline api/validator.go]: What do you mean by "If i do not set this value in yaml it must set system-default." ?

File: api/validator.go (hunk 2 above)
Comment (tamalsaha) [inline api/validator.go]: Need to relax this check to allow port not defined for HTTP case,

File: api/validator.go (hunk 4 above)
Comment (tamalsaha) [inline api/validator.go]: We can do this. But user can still do whatever they want in YAML. I believe the marshaler just ignores unknown fields.

File: api/validator.go (hunk 2 above)
Comment (tamalsaha) [inline api/validator.go]: Fixed.

File: api/validator.go (hunk 4 above)
Comment (tamalsaha) [inline api/validator.go]: Added.

File: api/port.go [hunk 6]
```diff
@@ -0,0 +1,98 @@
...
+func (r Ingress) PortMappings(cloudProvider string) (map[int]Target, error) {
+	mappings := make(map[int]Target)
+
+	usesHTTPRule := false
+	for _, rule := range r.Spec.Rules {
+		if rule.HTTP != nil {
+			usesHTTPRule = true
+			if _, foundTLS := r.FindTLSSecret(rule.Host); foundTLS && !rule.HTTP.NoSSL {
+				if port := rule.HTTP.Port.IntValue(); port > 0 {
+					mappings[port] = Target{PodPort: port, NodePort: rule.TCP.NodePort.IntValue()}
```
Comment (sadlil) [inline api/port.go]: `rule.HTTP.NodePort.IntValue()` isn't it?

Comment (tamalsaha) [inline api/port.go]: yes

**Comment by tamalsaha** [comment]
@sadlil , I have updated the validation logic (last commit).

**Pull request merged**